Rules system optimization

nobrakal opened the pull request:
Hi,

I have been working recently with fusion, and it made me think about some optimizations that can be made to the current system for `Algebra.Graph`.

This:

1. Change to good producers `vertices`, `edges` (ie. they are now expressed directly with `buildR`)... Meaning that `foldg e v o c . vertices xs` is optimized to a single list traversal _without a custom a rule_.

2. `bind` is also a good producer, so I added its `buildR` definition directly. 

3. `star` is now (in addition to be a good producer of graph), a good consumer of lists (meaning that `star . map f` is also optimized`).

4. The `foldr1Safe` rule is now unneeded, I removed it.

5. I added `INLINE` pragmas to functions expressed with `foldg`.

All the previous tests are passing, and the benchmarking suite is happy.

### Questions

I have some questions:

* Can we expose `buildR` ? Maybe moving it in `Algebra.Graph.Internal`, but exposing such a function allows the end-user to define its own function using `buildR` and benefit from fusion using `foldg/buildR`.

* The whole thing needs documentation. I like the "good producer" / "good consumer" vocabulary (stolen from the GHC manual). A good consumer is a function that can be fused with a good producer, and a good producer is a function that can be fused with a good consumer. Then we can tag all concerned functions.

* Because of the use of `buildR`, we _need_ rewrite rules to be enabled (either, some `foldg Empty Vertex Overlay Connect` will not be optimized). There is no problem for the library itself, but we should add a warning about it somewhere.

snowleopard commented:
@nobrakal Thanks for the PR! I like the overall goal: introducing the vocabulary of "good" producers and consumers and making sure that all exported functions are as good as possible!

It would, of course, be nice if users could write their own good producers and consumers.

My concern is that users who don't know (and perhaps don't want to know) about rewrite rules will be exposed to them if we start exporting functions like `buildR`. It would be great if the user could write simple code like `overlays . map vertex` and get an efficient implementation (i.e. a good producer).

I would have hoped for the notion of "goodness" to be composable, i.e. if `map` preservers goodness in the world of lists, and `overlays` is a good consumer of lists and a good producer of graphs, then their composition is automatically a good consumer of lists and a good producer of graphs. Is there any hope for achieving something like this? 

If we make goodness compositional we will never run out of goodness! :)

snowleopard commented:
P.S.: I guess what I'm after is a "good category" :-) The identity function `id` is hopefully both a good consumer and producer, and I'd like `.` to preserve goodness too!

nobrakal commented:
I added some comments, many functions are good producers and good consumers :D (`biclique` even ended to be a good consumer of _both_ its arguments and a good producer !)

> My concern is that users who don't know (and perhaps don't want to know) about rewrite rules ...

That why I propose to export it from `Algebra.Graph.Internal` (like `build` is exported from `GHC.Base` for lists). Then only people that are searching for it will find it :)

> I would have hoped for the notion of "goodness" to be composable ...

 I think that, strictly speaking, the notion of "goodness" is composable. The only problem here for me is that `overlays` is a good consumer of lists, and not of list _of graphs_. An other way to say is: because `buildR` abstracts all the constructors of graphs, thus any good producer should be expressed "in one go". For `overlays . map vertex` to be a good producer, `overlays` should know the previous `map vertex` to abstract the `Vertex` constructor.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -209,6 +213,7 @@ not under our control. We therefore choose the safe and more explicit path of
 creating our own intermediate functions for guiding rewrite rules when needed.
 -}
 
+-- | 'fmap' is a good consumer and a good producer.
```
Since these comments are already attached to specific functions, I suggest writing simply "A good consumer and producer", "A good consumer of both arguments", etc.

nobrakal replied:
For class functions, I think its worth mentioning which function is a good consumer/producer.

Otherwise, you are right, I will change this.

snowleopard commented:
> I added some comments, many functions are good producers and good consumers :D (`biclique` even ended to be a good consumer of both its arguments and a good producer !)

@nobrakal Great! :)

> That why I propose to export it from `Algebra.Graph.Internal` (like `build` is exported from `GHC.Base` for lists). Then only people that are searching for it will find it :)

OK, let's do that. I'd prefer to export it with a simpler name `build` (instead of `buildR`).

> The only problem here for me is that overlays is a good consumer of lists, and not of list of graphs. 

Oh, I see. That's more tricky than I thought -- there are layers of goodness!

nobrakal commented:
> OK, let's do that. I'd prefer to export it with a simpler name build or buildg (instead of buildR).

I said it before trying, but `buildR` needs to know the graphs constructors, so we cannot put it in `Algebra.Graph.Internal`.

snowleopard commented:
@nobrakal In this case, let's export it as `buildg` from `Algebra.Graph`.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1285,29 +1288,37 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
+-- Functions expressed with 'buildg' are good producers.
+--
+-- @
+-- buildg g = g 'Empty' 'Vertex' 'Overlay' 'Connect'
```
Could you add some more examples?

I guess you could add examples like `empty = buildg (\e _ _ _ -> e)` etc.

nobrakal replied:
Done, I also added corresponding tests

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1285,29 +1288,37 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
 * The "bindR/bindR" rule optimises compositions of multiple bindR's.
 -}
 
+-- | A function abstracting graphs constructors.
```
I think it's worth providing some more explanation. For example, pointing out that it's actually the type of `foldg`, right? I guess it is also related dto the Boehm-Berarducci encoding of algebraic graphs:

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph-Fold.html

Note, however, that `Fold` will be deleted in the next release... Maybe I shouldn't have done that :)

nobrakal replied:
> I guess it is also related dto the Boehm-Berarducci encoding of algebraic graphs

Ah, yes indeed. If you are going to remove it, I don't know if it worth adding a reference.

I am not fond of "this is foldg's type", because, for me, it hides the use-case of building a structure (ie, one can express `vertices` with `buildg`).

nobrakal commented:
I addressed your comments. More generally, I don't know the amount of documentation we need to add... There are many things to say :)

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1187,68 +1288,76 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
-buildR :: Foldg a -> Graph a
-buildR g = g Empty Vertex Overlay Connect
-{-# INLINE [1] buildR #-}
+-- | Replace abstracted graphs constructors by concrete ones.
```
```suggestion
-- | Build a graph given an interpretation of the four graph construction primitives 'empty',
-- 'vertex', 'overlay' and 'connect', in this order. See examples for further clarification.
```

snowleopard commented:
@nobrakal What about removing the `Foldg` synonym? I've suggested a comment that might work if you simply inline `Foldg` into the definition of `buildg`.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -47,7 +47,10 @@ module Algebra.Graph (
+    Context (..), context,
+
+    -- * Deforestation
+    buildg
```
I suggest placing it into the same category as `foldg`: these two functions are related and it would be nice if they appear in the documentation together.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1187,68 +1288,69 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
 -- Fuse composeR's. This occurs when two adjacent 'bindR' were rewritted into
--- their buildR form.
+-- their buildg form.
 "bindR/bindR" forall c f g.
```
Why is this rule called `bindR/bindR`? I guess it should be `composeR/composeR`.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -268,13 +279,15 @@ apR :: Graph (a -> b) -> Graph a -> Graph b
 bindR :: Graph a -> (a -> Graph b) -> Graph b
-bindR g f = foldg Empty f Overlay Connect g
-{-# INLINE [0] bindR #-}
+bindR g f = buildg $ \e v o c ->
```
Do we still need `bindR`? I couldn't find any remaining rules for it.

nobrakal replied:
Indeed we can remove them, the same apply to `apR` and `fmapR`, they were leftovers.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1187,68 +1288,69 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
+{-# INLINE [1] composeR #-}
 
 matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
 matchR e v p = \x -> if p x then v x else e
```
`matchR` doesn't seem to be used in the LHS of a rewrite rule either.

nobrakal replied:
You can find it in the `"graph/induce"` rule

snowleopard replied:
Indeed, thanks!

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -47,7 +47,7 @@ module Algebra.Graph (
 
     -- * Context
-    Context (..), context
+    Context (..), context,
```
Looks like an accidental change.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -173,12 +173,16 @@ compatible with 'overlay' and 'connect' operations:
              | Overlay (Graph a) (Graph a)
              | Connect (Graph a) (Graph a)
-             deriving (Show, Generic)
+             deriving (Generic)
```
I think it's best to continue using derived `Show`. Performance is not really that important for `Show` because it is not supposed to be used for large graphs anyway. Using derived `Show` allows us to avoid the extra complexity that comes with the custom implementation.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -232,9 +241,11 @@ instance Num a => Num (Graph a) where
     abs         = id
     negate      = id
 
+-- | `==` is a good consumer of both arguments.
```
You use backticks in a few places and I think Haddock doesn't support them -- please double-check the generated docs. I think we need to switch to quotes `'`.

nobrakal replied:
Backquotes seem to be the syntax for infix operators. This is producing a good output for me.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -423,21 +442,28 @@ overlays = fromMaybe empty . foldr1Safe overlay
+--
 -- @
--- foldg 'empty' 'vertex'        'overlay' 'connect'        == id
+-- foldg 'empty' 'vertex'        'overlay' 'connect'       == id
```
I think the alignment was fine in the Haddock:

https://hackage.haskell.org/package/algebraic-graphs-0.4/docs/Algebra-Graph.html#v:foldg

Are these changes accidental?

nobrakal replied:
They are not, the alignment was buggy on master for me. Could you check? Maybe my haddock has a problem.

snowleopard replied:
Hmm, that's strange, let me check.

snowleopard commented:
@nobrakal Apologies for taking so long to review and merge. I've left a few more comments/questions.

Is it possible to test the guarantees of being a good producer/consumer? I'm worried that we might accidentally break these guarantees at some point later without realising it.

nobrakal commented:
> Is it possible to test the guarantees of being a good producer/consumer? I'm worried that we might accidentally break these guarantees at some point later without realising it.

We can a write a `inspection-testing` test for each function indeed, but this will make a huge testsuite

snowleopard commented:
@nobrakal I would very much prefer to include tests for such documentation claims, because they are very likely to break otherwise.

Wouldn't one test per claim be sufficient?

snowleopard commented:
By the way, while we're tweaking rewrite rules, can we also remove this version constraint?

`inspection-testing >= 0.4 && < 0.4.2`

nobrakal commented:
Hi there, sorry for taking so long to answer.

> Wouldn't one test per claim be sufficient?

Indeed, but it will need more than 50 hand-made tests (the only test I see is to manually make the fusion and compare using `inspection-testing` for each claim). I can do the work, but I prefer to be sure we need it :)
Note that being a good consumer is just "consume the graph argument using `foldg`" and being a good producer is just "abstracting graphs constructors with `buildg`". They are pretty simple requirements but I don't see an easy way to test them.

> By the way, while we're tweaking rewrite rules, can we also remove this version constraint?

We just need to avoid version 0.4.2 which had a bug.

snowleopard commented:
@nobrakal Yes, adding 50 tests manually doesn't sound like fun. However, for every function in question, we already have a bunch of tests, so adding 1-2 new tests doesn't seem like a big deal. Thorough testing is pretty much always tedious, but it does bring benefits in the long term.

I can easily imagine myself accepting a PR two years from now that accidentally breaks the good consumer/producer property. A test would save the day :)

nobrakal commented:
@snowleopard Ok, I will add them.

I just rewrite old tests into more systematic ones. Do they seems OK to you ? If yes, I will follow with all the possible functions :)

By the way, I just hit a little problem with `<*>`, I will try to figure why it is not a good consumer of its second argument.

snowleopard commented:
@nobrakal Great, thank you! Yes, the tests look good.

> By the way, I just hit a little problem with `<*>`, I will try to figure why it is not a good consumer of its second argument.

Aha! So, tests are already proving useful :-)

nobrakal commented:
I added some tests for "simple" functions, I will investigate others as soon as possible :)

snowleopard commented:
@nobrakal Glad to see this PR is back to life! :-)

After having looked at the tests, I'm wondering: is it possible to get rid of duplication of many of the definitions from `Algebra.Graph`? Can we express the notions of "good producer" and "good consumer" in a less mechanical manner?

The README of the `inspection-testing` library lists the following as one of the features:

> checking the absence of a certain type (useful in the context of list or stream fusion)

So, what about checking that `size` is a good consumer and `biclique` is a good producer by applying one to another and testing that no `Graph` constructors remain?

nobrakal commented:
Yes, indeed you are right, I just tested and something like

```Haskell
overlaysC :: Build (Graph a) -> Graph a
overlaysC xs = overlays (build xs)

inspect $ 'overlaysC `hasNoType` ''[]
```

Works like a charm :)

There is two little problems:

1. This will not work with graph homomorphisms (the graph type will appear anyway since it is a good producer and a good consumer).
2. `foldg e v o c (overlays xs)` for example, or others functions using graph internally cannot be tested this way, since the graph type appears inside the function even after fusion.

I think I can add `hasNoType` where possible, and use old ugly duplication where it is not possible. Does it sounds good to you ?

nobrakal commented:
As I had some time, I rewrote concerned tests using your suggestion, the commit can be reverted without too many trouble I think.
I also added more tests (including the one for `hasEdge` !). We only now need to ensure that all claims are correctly tested, but I think the biggest part of the work is now made.

snowleopard commented:
@nobrakal This is great, the tests look much better!

For functions that consume and produce graphs, e.g. `transpose`, perhaps we could check that their composition with `buildg` and `foldg` has no `Graph` constructors? I think if `transpose` is indeed a good consumer and a good producer, then `\e v o c g -> foldg e v o c (transpose (buildg g))` should have no graph constructors left, i.e. the result should be just `g e v o (flip c)` :-)

snowleopard commented:
In fact, I think this "sandwich" approach works for any function that claims to be both a good consumer of X and a good producer of Y: i.e. we can just sandwich it between a producer of X and a consumer of Y and see what happens.

nobrakal commented:
@snowleopard  You are right, the sandwich approach would work. I can only argue that it is merging two tests for two different things into a single test, but I don't know if this is a problem.

snowleopard commented:
> I can only argue that it is merging two tests for two different things into a single test, but I don't know if this is a problem.

Intuitively, only a function that is both good consumer and a good producer can pass the sandwich check. Do you agree? I don't see a simple formal argument :-)

nobrakal commented:
Ok, I agree this is simpler anyway :)

Just a question about `induceJust`, should I add the same set of rules than `induce` to mimic its fusion (since it is not a good producer naturally) ?

snowleopard commented:
@nobrakal Can't `induce` and `induceJust` be turned into good producers by using `buildg`? I think it's possible, although we might need to add `Maybe` to pattern match against `Nothing` instead of `Empty`. Do you see what I mean?

nobrakal commented:
@snowleopard I am not sure why, I was sure to have tested this solution and rejected it for efficiency reason. Anyway, I just gave it a try, and all seems to work :) I will wait to see if the benchmarking suite is showing terrible results.

nobrakal commented:
`removeVertex` testing `induce is perfect!

Inspecting the benchmarking results, I just discovered a little flow I introduced that was dropping the performances of `edges` and corrected it in the last commit.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -232,49 +234,51 @@ instance Num a => Num (Graph a) where
 -- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
 eqIntR :: Graph Int -> Graph Int -> Bool
-eqIntR x y = toAdjacencyIntMap x == toAdjacencyIntMap y
+eqIntR = \x y -> toAdjacencyIntMap x == toAdjacencyIntMap y
```
Is this change accidental? If not, what does it achieve?

nobrakal replied:
This was a little trick when I compared directly the definitions of functions and expected form. This is not needed anymore.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -232,49 +234,51 @@ instance Num a => Num (Graph a) where
 -- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
 ordIntR :: Graph Int -> Graph Int -> Ordering
-ordIntR x y = compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
+ordIntR = \x y -> compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
```
Same question here.

nobrakal replied:
Same answer :)

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -232,49 +234,51 @@ instance Num a => Num (Graph a) where
+ordIntR = \x y -> compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
+{-# INLINE ordIntR #-}
 
+-- | `<*>` is a good consumer of its first agument and producer.
```
Is it possible to make it a good consumer of both arguments? If not, is there a simple argument why this can't be done?

snowleopard replied:
(If there is such an argument, I think it'd be useful to add it above as a comment.)

nobrakal replied:
There is no arguments, I don't get it. Inspecting core, the second argument is hidden in a local variable until last phase, preventing the rule to fire... I spent some time without success, so I will add a comment for now.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -634,17 +673,20 @@ edgeCountIntR = AIM.edgeCount . toAdjacencyIntMap
 vertexList :: Ord a => Graph a -> [a]
 vertexList = Set.toAscList . vertexSet
-{-# INLINE [1] vertexList #-}
+{-# INLINE [2] vertexList #-}
```
I think `vertexList` and `edgeList` are good consumers too.

nobrakal replied:
Indeed, I even tested it ^^

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -715,44 +768,58 @@ adjacencyList = AM.adjacencyList . toAdjacencyMap
 -- Complexity: /O(L)/ time, memory and size, where /L/ is the length of the
 -- given list.
 --
+-- Good producer.
```
Can `path` and `circuit` be made good consumers of lists?

nobrakal replied:
The use of direct pattern-matching is the problem here. You will have to write the two functions with `foldr` for them to be a good consumer of lists. I don't see right now a clear way to do that while still being efficient and elegant.

snowleopard replied:
OK, your reply looks like a good TODO comment to me :)

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1051,6 +1153,8 @@ induceJust = foldg Empty (maybe Empty Vertex) (k Overlay) (k Connect)
 -- Complexity: the function performs /O(s)/ graph comparisons. It is guaranteed
 -- that the size of the result does not exceed the size of the given expression.
 --
+-- Good consumer.
```
Let's make it a good producer too (seems straightforward).

nobrakal replied:
This is not straightforward: when you abstract your constructors using `buildg` you loose the `Eq` instance required by `simple` (intuitively, you could destruct your graph to a value which does not have an `Eq` instance)

snowleopard replied:
Hmm, I see, that's interesting. Worth documenting this intuition!

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1086,6 +1191,8 @@ simple op x y
 -- quadratic, i.e. /m = O(m1 * m2)/, but the algebraic representation requires
 -- only /O(m1 + m2)/ operations to list them.
 --
+-- Good consumer of both arguments.
```
Does it happen to also be a good producer? I'd hope that `overlays` of `biclique`s can fuse nicely.

snowleopard replied:
(Perhaps the list of bicliques does disappear, but not the bicliques themselves...)

nobrakal replied:
Indeed, it is a good producer :)

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1202,81 +1310,63 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
-buildR :: Foldg a -> Graph a
-buildR g = g Empty Vertex Overlay Connect
-{-# INLINE [1] buildR #-}
+-- | Build a graph given an interpretation of the four graph construction primitives 'empty',
```
Let's move this definition up, after `foldg`, to match the export order (just for ease of navigation).

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -1202,81 +1310,63 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
+    composeR (composeR c f) g = composeR c (f . g)
 
--- Rewrite identity (which can appear in the rewriting of bindR) to a much efficient one
+-- Rewrite identity (which can appear in the inlining of 'buildg') to a much efficient one.
```
much -> more/much more

snowleopard commented on src/Algebra/Graph/Internal.hs:
```diff
@@ -108,14 +108,7 @@ connectFoci x y = Focus (ok x || ok y) (xs <> is y) (os x <> ys) (vs x <> vs y)
 -- | A safe version of 'foldr1'.
 foldr1Safe :: (a -> a -> a) -> [a] -> Maybe a
```
I think this function can actually be called `sconcat`, i.e. concatenating a list of semigroup values :)

nobrakal replied:
Indeed :) Do you not fear a name clash with the actual `Data.Semigroup.sconcat` ?

snowleopard replied:
Ah, good point. We do often import `Data.Semigroup`, so the name clash may become annoying. OK, let's keep `foldr1Safe` for now.

snowleopard commented on src/Algebra/Graph.hs:
```diff
@@ -381,25 +388,31 @@ connect = Connect
 -- 'vertexSet'   . vertices == Set.'Set.fromList'
 -- @
 vertices :: [a] -> Graph a
-vertices = overlays . map vertex
```
Any hope (perhaps, just in future) to be able to implement `vertices` as `overlays . map vertex` preserving the goodness? :)

nobrakal replied:
Sadly, goodness behave well with composition only when types agree.
`overlays . map vertex` will still be a good consumer of lists, but not a good producer, since `overlays` is only a good consumer of lists, and not of lists _of graphs_.
To say it more clearly, if we abstract constructors for the whole `vertices`, we need to reuse the same abstract constructors in `overlays`.

snowleopard replied:
I see. I think you have already explained this to me once :) Could you add it as a comment, perhaps, with a TODO (that we'd like to simplify this to `overlays . map vertex`).

snowleopard commented on test/Algebra/Graph/Test/RewriteRules.hs:
```diff
@@ -13,111 +14,352 @@ module Algebra.Graph.Test.RewriteRules where
 
-inspect $ 'starT1 === 'starTR
+-- simplify
+simple :: Eq g => (g -> g -> g) -> g -> g -> g
```
I guess if `simplify` becomes a good producer we won't need this re-definition.

snowleopard commented on test/Algebra/Graph/Test/RewriteRules.hs:
```diff
@@ -13,111 +14,352 @@ module Algebra.Graph.Test.RewriteRules where
+-- compose
+composeC, composeCR :: Ord a => Buildg a -> Buildg a -> Graph a
+composeC  x y = compose (buildg x) (buildg y)
+composeCR x y = overlays
```
Ouch. It's a bit sad we need to copy this.

nobrakal replied:
I agree but I don't see a better option here...

snowleopard commented:
@nobrakal I had a thorough look through the changes and left a bunch of comments, most of which should be easy to address. Hope we can merge this soon!

nobrakal commented:
Thank you very much for your detailed review ! I addressed "simple" issues and will come back soon to more difficult ones ^^

snowleopard commented:
@nobrakal Thanks for addressing my comments! 

By the way, it looks like there are some merge conflicts. Please rebase.

nobrakal commented:
@snowleopard I do not understand: it seems I am up-to-date with `master`. A local merge passes without problem on my machine, and even github seems happy with the PR.

snowleopard commented:
@nobrakal Oops, looks like a GitHub bug. I see 

> This branch cannot be rebased due to conflicts
Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.

When "Rebase and merge" is selected. If I switch to "Squash and merge", which is more appropriate in this case anyway, everything gets green!

snowleopard commented:
@nobrakal OK, I think we can merge this if you are ready!

nobrakal commented:
I just found a new good consumer! Except this, all seems green for a merge :)

snowleopard commented:
@nobrakal Merged! Many thanks for this awesome improvement :-)

fawaz990 commented:
ابغا أصير مشرف برامج ومواقع اتواصلو معي
